parsers/cpreprocessor.c

```diff
@@ -1244,54 +1300,380 @@ static void saveIgnoreToken(const char * ignoreToken, const char * param_name)
+	}
+
+	while(*c && isspacetab(*c))
+		c++;
+
+	info->replacements = NULL;
+
+
+	if(*c == '=')
+	{
+		CXX_DEBUG_PRINT("Macro has a replacement part");
+
+		// have replacement part
+		c++;
+		
+		cppMacroReplacementPartInfo * lastReplacement = NULL;
+		int nextParameterReplacementFlags = 0;
+
+#define ADD_REPLACEMENT_NEW_PART(part) \
+		do { \
+			if(lastReplacement) \
+				lastReplacement->next = part; \
+			else \
+				info->replacements = part; \
+			lastReplacement = part; \
+		} while(0)
+
+#define ADD_CONSTANT_REPLACEMENT_NEW_PART(start,len) \
+		do { \
+			cppMacroReplacementPartInfo * rep = (cppMacroReplacementPartInfo *)eMalloc(sizeof(cppMacroReplacementPartInfo)); \
+			rep->parameterIndex = -1; \
+			rep->flags = 0; \
+			rep->constant = vStringNew(); \
+			vStringNCatS(rep->constant,start,len); \
+			rep->next = NULL; \
+			CXX_DEBUG_PRINT("Constant replacement part: '%s'",vStringValue(rep->constant)); \
+			ADD_REPLACEMENT_NEW_PART(rep); \
+		} while(0)
+
+#define ADD_CONSTANT_REPLACEMENT(start,len) \
+		do { \
+			if(lastReplacement && (lastReplacement->parameterIndex == -1)) \
+			{ \
+				vStringNCatS(lastReplacement->constant,start,len); \
+				CXX_DEBUG_PRINT("Constant replacement part changed: '%s'",vStringValue(lastReplacement->constant)); \
+			} else { \
+				ADD_CONSTANT_REPLACEMENT_NEW_PART(start,len); \
+			} \
+		} while(0)
+
+		// parse replacements
+		const char * begin = c;
+		
+		while(*c)
+		{
+			if(isalpha(*c) || (*c == '_'))
+			{
+				if(c > begin)
+					ADD_CONSTANT_REPLACEMENT(begin,c - begin);
+
+				const char * tokenBegin = c;
+
+				while(*c && (isalnum(*c) || (*c == '_')))
+					c++;
+
+				// check if it is a parameter
+				int tokenLen = c - tokenBegin;
+
+#ifdef CXX_DO_DEBUGGING
+				char tmp = *c;
+				*(char *)c = 0;
+				CXX_DEBUG_PRINT("Check token '%s'",tokenBegin);
+				*(char *)c = tmp;
+#endif
+
+				bool bIsVarArg = strncmp(tokenBegin,"__VA_ARGS__",tokenLen) == 0;
+
+				int i = 0;
+				for(;i<iParamCount;i++)
+				{
+					int paramLen = paramEnd[i] - paramBegin[i];
+					
+					if(
+							(bIsVarArg && (paramLen == 3) && (strncmp(paramBegin[i],"...",3) == 0)) ||
+							((!bIsVarArg) && (paramLen == tokenLen) && (strncmp(paramBegin[i],tokenBegin,paramLen) == 0))
+						)
+					{
+						// parameter!
+						cppMacroReplacementPartInfo * rep = (cppMacroReplacementPartInfo *)eMalloc(sizeof(cppMacroReplacementPartInfo));
+						rep->parameterIndex = i;
+						rep->flags = nextParameterReplacementFlags | (bIsVarArg ? CPP_MACRO_REPLACEMENT_FLAG_VARARGS : 0);
+						rep->constant = NULL;
+						rep->next = NULL;
+
+						nextParameterReplacementFlags = 0;
+
+						CXX_DEBUG_PRINT("Parameter replacement part: %d (vararg %d)",i,bIsVarArg);
+
+						ADD_REPLACEMENT_NEW_PART(rep);
+						break;
+					}
+				}
+				
+				if(i >= iParamCount)
+				{
+					// no parameter found
+					ADD_CONSTANT_REPLACEMENT(tokenBegin,tokenLen);
+				}
+				
+				begin = c;
+				continue;
```

Comment:
I don't like this `continue` and the others in this loop, that, unless I'm mistaken, don't give anything over a simple `if() else if() else if() else` construction